src/main/scala/diplomacy/Parameters.scala

```diff
@@ -322,3 +322,44 @@ trait DirectedBuffers[T] {
   def copyOut(x: BufferParams): T
   def copyInOut(x: BufferParams): T
 }
+
+trait IdMapEntry {
+  def name: String
+  def from: IdRange
+  def to: IdRange
+  def isCache: Boolean
+  def requestFifo: Boolean
+  def pretty(fmt: String) =
+    if (from ne to) { // if the subclass uses the same reference for both from and to, assume its format string has an arity of 5
+      fmt.format(to.start, to.end, from.start, from.end, s""""$name"""", if (isCache) " [CACHE]" else "", if (requestFifo) " [FIFO]" else "")
+    } else {
+      fmt.format(from.start, from.end, s""""$name"""", if (isCache) " [CACHE]" else "", if (requestFifo) " [FIFO]" else "")
+    }
+}
+
+abstract class IdMap[T <: IdMapEntry] {
+  protected val fmt: String
+  val mapping: Seq[T]
+  def pretty: String = mapping.map(_.pretty(fmt)).mkString(",\n")
+}
+
+class IdRangeSerial (val start: Int, val end: Int)
+object IdRangeSerial {
+  def apply(i: IdRange): IdRangeSerial = {
+    new IdRangeSerial(i.start, i.end)
+  }
+}
```

Comment:
```suggestion
case class IdRangeSerial (val start: Int, val end: Int)

object IdRangeSerial {
  def apply(i: IdRange): IdRangeSerial = IdRangeSerial(i.start, i.end)
}
```

This makes sense as a case class to me

Reply:
though now that I wrote that, i wonder if it will conflict with the IdRangeSerial.apply() method. 
@jackkoenig thoughts?

I guess worth asking is there a reason you did not want to make this a case class?

Reply:
im going to resolve this comment since I think it's irrelevant given my more recent comment